node/core/backing/src/lib.rs

```diff
@@ -786,626 +741,1277 @@ async fn validate_and_make_available(
+	};
+
+	for deactivated in update.deactivated {
+		state.per_leaf.remove(&deactivated);
+		state.implicit_view.deactivate_leaf(deactivated);
+	}
+
+	// clean up `per_relay_parent` according to ancestry
+	// of leaves. we do this so we can clean up candidates right after
+	// as a result.
+	//
+	// when prospective parachains are disabled, the implicit view is empty,
+	// which means we'll clean up everything. This is correct.
+	for relay_parent in state.implicit_view.all_allowed_relay_parents() {
+		state.per_relay_parent.remove(relay_parent);
+	}
+
+	// clean up `per_candidate` according to which relay-parents
+	// are known.
+	//
+	// when prospective parachains are disabled, we clean up all candidates
+	// because we've cleaned up all relay parents. this is correct.
+	state
+		.per_candidate
+		.retain(|_, pc| state.per_relay_parent.contains_key(&pc.relay_parent));
+
+	// Get relay parents which might be fresh but might be known already
+	// that are explicit or implicit from the new active leaf.
+	let fresh_relay_parents = match res {
+		None => return Ok(()),
+		Some((leaf, LeafHasProspectiveParachains::Disabled)) => {
+			// defensive in this case - for enabled, this manifests as an error.
+			if state.per_leaf.contains_key(&leaf.hash) {
+				return Ok(())
+			}
+
+			state.per_leaf.insert(
+				leaf.hash,
+				ActiveLeafState {
+					prospective_parachains_mode: ProspectiveParachainsMode::Disabled,
+					// This is empty because the only allowed relay-parent and depth
+					// when prospective parachains are disabled is the leaf hash and 0,
+					// respectively. We've just learned about the leaf hash, so we cannot
+					// have any candidates seconded with it as a relay-parent yet.
+					seconded_at_depth: BTreeMap::new(),
+				},
+			);
+
+			vec![leaf.hash]
+		},
+		Some((leaf, LeafHasProspectiveParachains::Enabled(Ok(_)))) => {
+			let fresh_relay_parents =
+				state.implicit_view.known_allowed_relay_parents_under(&leaf.hash, None);
+
+			// At this point, all candidates outside of the implicit view
+			// have been cleaned up. For all which remain, which we've seconded,
+			// we ask the prospective parachains subsystem where they land in the fragment
+			// tree for the given active leaf. This comprises our `seconded_at_depth`.
+
+			let remaining_seconded = state
+				.per_candidate
+				.iter()
+				.filter(|(_, cd)| cd.seconded_locally)
+				.map(|(c_hash, cd)| (*c_hash, cd.para_id));
+
+			// one-to-one correspondence to remaining_seconded
+			let mut membership_answers = FuturesOrdered::new();
+
+			for (candidate_hash, para_id) in remaining_seconded {
+				let (tx, rx) = oneshot::channel();
+				membership_answers.push(rx.map_ok(move |membership| (candidate_hash, membership)));
+
+				ctx.send_message(ProspectiveParachainsMessage::GetTreeMembership(
+					para_id,
+					candidate_hash,
+					tx,
+				))
+				.await;
+			}
+
+			let mut seconded_at_depth = BTreeMap::new();
+			for response in membership_answers.next().await {
```

Comment:
```suggestion
			while let Some(response) = membership_answers.next().await {
```

Reply:
the existing code here only polls it once, I think, and compiles because `Option<T>` implements `IntoIterator<T>`